File: CoreAutomata/AutomataApi.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading;

namespace CoreAutomata
{
    /// <summary>
    /// <see cref="AutomataApi"/> is the part that connects the Script project and the Android project.
    /// The Script project works as if the dimensions of the screen are as set in <see cref="GameAreaManager.ScriptDimension"/>.
    /// The Android project works in original screen pixel size.
    /// It is the work of <see cref="AutomataApi"/> to Transform the values (<see cref="Region"/>, <see cref="Location"/>) between these coordinate spaces such that both the projects are unaware of any transformation happening.
    /// Use the Extension method variation of the methods wherever possible.
    /// </summary>
    public static class AutomataApi
    {
        static IPlatformImpl _platformImpl;
        static IGestureService _gesture;

        static AutomataApi()
        {
            DebugMsgReceived += Msg =>
            {
                Msg = $"DBG: {Msg}";

                Console.WriteLine(Msg);

                Toast(Msg);
            };
        }

        public static void RegisterPlatform(IPlatformImpl Impl)
        {
            _platformImpl = Impl;
        }

        public static void RegisterGestures(IGestureService GestureService)
        {
            _gesture = GestureService;
        }

        /// <summary>
        /// The default matching similarity used unless explicitly specified in methods like <see cref="Exists"/> and <see cref="WaitVanish"/>.
        /// </summary>
        public static double MinSimilarity { get; set; } = 0.8;

        /// <summary>
        /// The default duration for <see cref="Highlight"/> method.
        /// </summary>
        public static TimeSpan DefaultHighlightTimeout { get; set; } = TimeSpan.FromSeconds(0.3);

        public static IPattern LoadPattern(Stream Stream, bool Mask = false)
        {
            return _platformImpl.LoadPattern(Stream, Mask);
        }

        public static void Wait(this TimeSpan TimeSpan)
        {
            Thread.Sleep(TimeSpan);
        }

        public static Region WindowRegion => _platformImpl.WindowRegion;

        public static void Click(this Location Location) => _gesture.Click(Location.Transform());

        static bool ExistsNow(Region Region, IPattern Image, double? Similarity)
        {
            var sshot = ScreenshotManager.GetScreenshot();

            if (Region != null)
            {
                if (_platformImpl.DebugMode)
                {
                    Region.Highlight();
                }

                sshot = sshot.Crop(Region.TransformToImage());
            }

            return sshot
                .IsMatch(Image, Similarity ?? MinSimilarity);
        }

        static readonly Stopwatch Stopwatch = new Stopwatch();

        static bool CheckConditionLoop(Func<bool> Condition, TimeSpan? Timeout)
        {
            Stopwatch.Restart();

            try
            {
                while (true)
                {
                    var scanStartTimestamp = Stopwatch.Elapsed;

                    if (Condition.Invoke())
                    {
                        return true;
                    }

                    if (Timeout == null || Stopwatch.Elapsed > Timeout)
                    {
                        break;
                    }

                    var scanInterval = TimeSpan.FromMilliseconds(1000 / ScanRate);
                    var elapsed = Stopwatch.Elapsed - scanStartTimestamp;
                    var timeToWait = scanInterval - elapsed;

                    if (timeToWait > TimeSpan.Zero)
                    {
                        Wait(timeToWait);
                    }
                }
            }
            finally
            {
                Stopwatch.Reset();
            }

            return false;
        }

        public static bool WaitVanish(this Region Region, IPattern Image, TimeSpan? Timeout = null, double? Similarity = null)
        {
            return CheckConditionLoop(() => !ExistsNow(Region, Image, Similarity), Timeout);
        }

        public static bool Exists(this Region Region, IPattern Image, TimeSpan? Timeout = null, double? Similarity = null)
        {
            return CheckConditionLoop(() => ExistsNow(Region, Image, Similarity), Timeout);
        }

        /// <summary>
        /// The number of times methods like <see cref="Exists"/> and <see cref="WaitVanish"/> check the screen in a second.
        /// </summary>
        public static double ScanRate { get; set; } = 3;

        public static IPattern GetPattern(this Region Region) => ScreenshotManager.GetScreenshot()
            .Crop(Region.TransformToImage())
            .Copy();

        public static void UseSameSnapIn(Action Action) => UseSameSnapIn(() =>
        {
            Action();
            return 0;
        });

        public static T UseSameSnapIn<T>(Func<T> Action)
        {
            ScreenshotManager.Snapshot();

            try
            {
                return Action();
            }
            finally
            {
                ScreenshotManager.UsePreviousSnap = false;
            }
        }

        public static void Swipe(Location Start, Location End) => _gesture.Swipe(Start.Transform(), End.Transform());

        public static IEnumerable<Match> FindAll(this Region Region, IPattern Pattern, double? Similarity = null)
        {
            var sshot = ScreenshotManager.GetScreenshot();

            if (Region != null)
            {
                if (_platformImpl.DebugMode)
                {
                    Region.Highlight();
                }

                sshot = sshot.Crop(Region.TransformToImage());
            }

            return sshot
                .FindMatches(Pattern, Similarity ?? MinSimilarity)
                .Select(M =>
                {
                    var region = M.TransformFromImage();

                    if (Region != null)
                    {
                        region.X += Region.X;
                        region.Y += Region.Y;
                    }
                    
                    return new Match(region, M.Score);
                });
        }

        public static IPattern GetResizableBlankPattern() => _platformImpl.GetResizableBlankPattern();

        public static void Toast(string Msg) => _platformImpl?.Toast(Msg);

        public static void ContinueClick(this Location Location, int Times)
        {
            _gesture.ContinueClick(Location.Transform(), Times);
        }

        public static void ShowMessageBox(string Title, string Message)
        {
            _platformImpl.MessageBox(Title, Message);
        }

        public static void Highlight(this Region Region, TimeSpan? Timeout = null)
        {
            _platformImpl.Highlight(Region.Transform(), Timeout ?? DefaultHighlightTimeout);
        }

        public static void SetStorageRootDir(string Dir)
        {
            _storageRootDir = Dir;
        }

        static string _storageRootDir;

        public static string StorageDir
        {
            get
            {
                var dir = Path.Combine(_storageRootDir, "Fate-Grand-Automata");

                if (!Directory.Exists(dir))
                {
                    Directory.CreateDirectory(dir);
                }

                return dir;
            }
        }

        [Conditional("DEBUG")]
        public static void WriteDebug(string Msg)
        {
            DebugMsgReceived?.Invoke(Msg);
        }

        public static event Action<string> DebugMsgReceived;
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/core/AutomataApi.kt
package com.mathewsachin.fategrandautomata.core

import com.mathewsachin.fategrandautomata.scripts.prefs.Preferences
import java.io.InputStream
import kotlin.math.min
import kotlin.time.Duration
import kotlin.time.TimeSource.Monotonic
import kotlin.time.milliseconds
import kotlin.time.seconds

/**
 * Checks if the stop button has been pressed.
 *
 * @throws ScriptAbortException if the button has been pressed
 */
fun checkExitRequested() {
    if (AutomataApi.exitRequested) {
        throw ScriptAbortException()
    }
}

/**
 * Clicks on the [Location].
 *
 * @param Times the amount of times to click
 */
fun Location.click(Times: Int = 1) {
    checkExitRequested()
    AutomataApi.GestureService?.click(this.transform(), Times)
}

/**
 * Checks if the [Region] contains the provided image.
 *
 * @param Image the image to look for
 * @param Timeout how long to search for before giving up
 * @param Similarity the minimum similarity for this search
 */
fun Region.exists(
    Image: IPattern,
    Timeout: Duration = Duration.ZERO,
    Similarity: Double = AutomataApi.MinSimilarity
): Boolean {
    checkExitRequested()
    return AutomataApi.checkConditionLoop(
        { AutomataApi.existsNow(this, Image, Similarity) },
        Timeout
    )
}

/**
 * Waits until the given image cannot be found in the [Region] anymore.
 *
 * @param Image the image to search for
 * @param Timeout how long to wait for before giving up
 * @param Similarity the minimum similarity for this search
 */
fun Region.waitVanish(
    Image: IPattern,
    Timeout: Duration,
    Similarity: Double = AutomataApi.MinSimilarity
): Boolean {
    checkExitRequested()
    return AutomataApi.checkConditionLoop(
        { !AutomataApi.existsNow(this, Image, Similarity) },
        Timeout
    )
}

/**
 * Clicks on the center of this Region.
 */
fun Region.click() = center.click()

/**
 * Gets the width and height in the form of a [Size] object.
 */
val IPattern.Size get() = Size(width, height)

/**
 * Adds borders around the [Region].
 *
 * @param Duration how long the borders should be displayed
 */
fun Region.highlight(Duration: Duration = 0.3.seconds) {
    checkExitRequested()
    AutomataApi.PlatformImpl?.highlight(this.transform(), Duration)
}

/**
 * Gets the image content of this Region.
 *
 * @return an [IPattern] object with the image data
 */
fun Region.getPattern(): IPattern? {
    return ScreenshotManager.getScreenshot()
        ?.crop(this.transformToImage())
        ?.copy()
}

/**
 * Searches for all occurrences of a given image in the [Region].
 *
 * @param Pattern the image to search for
 * @param Similarity the minimum similarity for this search
 *
 * @return a list of all matches in the form of [Match] objects
 */
fun Region.findAll(
    Pattern: IPattern,
    Similarity: Double = AutomataApi.MinSimilarity
): Sequence<Match> {
    var sshot = ScreenshotManager.getScreenshot()

    if (Preferences.DebugMode) {
        this.highlight()
    }

    sshot = sshot!!.crop(this.transformToImage())

    return sshot
        .findMatches(Pattern, Similarity)
        .map {
            checkExitRequested()

            var region = it.Region.transformFromImage()

            // convert the relative position in the region to the absolute position on the screen
            region += this.location

            Match(region, it.score)
        }
}

/**
 * Central class used for triggering gestures and image recognition.
 */
class AutomataApi {
    companion object {
        var PlatformImpl: IPlatformImpl? = null
        var GestureService: IGestureService? = null

        fun registerPlatform(Impl: IPlatformImpl) {
            PlatformImpl = Impl
        }

        fun registerGestures(Impl: IGestureService) {
            GestureService = Impl
        }

        /**
         * The default minimum similarity used for image comparisons.
         */
        const val MinSimilarity = 0.8

        fun loadPattern(Stream: InputStream): IPattern {
            return PlatformImpl!!.loadPattern(Stream)
        }

        fun getResizableBlankPattern(): IPattern {
            return PlatformImpl!!.getResizableBlankPattern()
        }

        /**
         * Wait for a given [Duration]. The wait is paused regularly to check if the stop button has
         * been pressed.
         */
        fun wait(Duration: Duration) {
            val epsilon = 1000L
            var left = Duration.toLongMilliseconds()

            // Sleeping this way allows quick exit if demanded by user
            while (left > 0) {
                checkExitRequested()

                val toSleep = min(epsilon, left)
                Thread.sleep(toSleep)
                left -= toSleep
            }
        }

        @Volatile
        var exitRequested = false

        val WindowRegion: Region get() = PlatformImpl!!.windowRegion

        /**
         * Shows a message box with the given title and message.
         */
        fun showMessageBox(Title: String, Message: String, Error: Exception? = null) {
            PlatformImpl?.messageBox(Title, Message, Error)
        }

        /**
         * Shows a toast with the given message.
         */
        fun toast(Message: String) {
            PlatformImpl?.toast(Message)
        }

        /**
         * Checks if the [Region] contains the provided image.
         *
         * @param Region the search region
         * @param Image the image to look for
         * @param Similarity the minimum similarity for this search
         */
        fun existsNow(
            Region: Region,
            Image: IPattern,
            Similarity: Double = MinSimilarity
        ): Boolean {
            checkExitRequested()

            var sshot = ScreenshotManager.getScreenshot()

            if (Preferences.DebugMode) {
                Region.highlight()
            }

            sshot = sshot?.crop(Region.transformToImage())

            return sshot?.isMatch(Image, Similarity) ?: false
        }

        /**
         * Repeats the invocation of the Condition until it returns `true` or until the timeout has
         * been reached.
         *
         * @param Condition a function with a [Boolean] return value
         * @param Timeout how long to wait for before giving up
         * @return `true` if the function returned `true` at some point, `false` if the timeout was
         * reached
         */
        fun checkConditionLoop(
            Condition: () -> Boolean,
            Timeout: Duration = Duration.ZERO
        ): Boolean {
            val endTimeMark = Monotonic.markNow() + Timeout

            while (true) {
                val scanStart = Monotonic.markNow()

                if (Condition.invoke()) {
                    return true
                }

                // check if we need to cancel because of timeout
                if (endTimeMark.hasPassedNow()) {
                    break
                }

                /* Wait a bit before checking again.
                   If invocationDuration is greater than the scanInterval, we don't wait. */
                val scanInterval = 330.milliseconds
                val timeToWait = scanInterval - scanStart.elapsedNow()

                if (timeToWait.isPositive()) {
                    wait(timeToWait)
                }
            }

            return false
        }

        /**
         * Takes a screenshot and caches it for the duration of the function invocation. This is
         * useful when you want to reuse the same screenshot for multiple image searches.
         *
         * @param Action a function to invoke which will use the cached screenshot
         */
        fun <T> useSameSnapIn(Action: () -> T): T {
            ScreenshotManager.snapshot()

            try {
                return Action()
            } finally {
                ScreenshotManager.usePreviousSnap = false
            }
        }

        /**
         * Swipes from one [Location] to another [Location].
         *
         * @param Start the [Location] where the swipe should start
         * @param End the [Location] where the swipe should end
         */
        fun swipe(Start: Location, End: Location) {
            GestureService?.swipe(Start.transform(), End.transform())
        }
    }
}